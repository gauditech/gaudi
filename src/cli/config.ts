import path from "path";

import { EngineConfig } from "@src/config";

// -------------------- Various configs, constants and defaults

/**
 * Time to wait before reporting resource watcher changes (in millis).
 *
 * This allows basic debouncing (eg. when creating/updating multiple files at once).
 */
export const RESOURCE_WATCH_DELAY = 500;

/** Returns path to prisma schema file */
export function getDbSchemaPath(config: EngineConfig): string {
  return `${config.gaudiFolder}/db/schema.prisma`;
}

// defaul Node options
const DEFAULT_NODE_OPTIONS = [
  // development node options - maybe we should allow disabling them in production?
  "--enable-source-maps",
  "--stack-trace-limit=30",
  "--stack-size=2048",
];

/**
 * Default node options
 *
 * Additional node options can be passed via NODE_OPTIONS env var which is included when running commands
 */
export function getDefaultNodeOptions(): string[] {
  return [...DEFAULT_NODE_OPTIONS];
}

/**
 * Paths to Gaudi scripts
 *
 * Target Gaudi scripts directly instead of via NPX because we cannot pass some node options through NPX (via --node-options")
 * See a list of allowed options here: https://nodejs.org/docs/latest-v16.x/api/cli.html#node_optionsoptions
 */
export const GAUDI_SCRIPTS = {
  ENGINE: path.join(__dirname, "../engine.js"),
  RUNTIME: path.join(__dirname, "../runtime/runtime.js"),
  POPULATOR: path.join(__dirname, "../populator/populator.js"),
};

/**
 * Add path to CLI's binary.
 *
 * We don't want to force users to be aware of our binaries which means that
 * they probably won't be available as global binaries thus we have to manually
 * target them in our own `node_modules`.
 */
export function appendBinPath(binName: string): string {
  return path.join("./node_modules/@gaudi/engine/node_modules/.bin/", binName);
}
