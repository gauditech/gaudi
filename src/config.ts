import dotenv from "dotenv";
import { ConnectionOptions } from "pg-connection-string";

import { parseConnectionString } from "./common/utils";

import { GAUDI_FOLDER_NAME } from "@src/const";

export type EngineConfig = {
  /** Path to Gaudi blueprint file */
  inputPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** Gaudi folder */
  gaudiFolder: string;
  dbConn: ConnectionOptions;
  embeddedPg: boolean;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(configPath?: string): EngineConfig {
  dotenv.config({ path: configPath });

  const inputPath = process.env.GAUDI_ENGINE_INPUT_PATH ?? "";
  const outputFolder = process.env.GAUDI_ENGINE_OUTPUT_PATH ?? ".";
  // gaudi folder's path should probably be determined by the position of (future) gaudi config file
  const gaudiFolder = `./${GAUDI_FOLDER_NAME}`;

  const dbConn = parseConnectionString(process.env.GAUDI_DATABASE_URL);
  const embeddedPg = process.env.GAUDI_EMBEDDED_POSTGRESQL_ENABLED?.toLowerCase() === "true";

  return { inputPath, outputFolder, gaudiFolder, dbConn, embeddedPg };
}
