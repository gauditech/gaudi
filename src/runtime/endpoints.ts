import express, { Request } from "express";
import knex from "knex";

const app = express();
const port = 3002;

import {
  PathParam,
  buildEndpointPath,
  queryableFromEndpointTargets,
  selectToSelectable,
} from "@src/builder/query";
import { Definition, QueryDef, QueryDefPath, TargetDef } from "@src/types/definition";

function validatePathParam(param: PathParam["params"][number], val: string): PathVariable {
  switch (param.type) {
    case "integer":
      return { name: param.name, type: param.type, value: parseInt(val, 10) };
    case "text":
      return { name: param.name, type: param.type, value: val };
  }
}

type PathVariable =
  | { name: string; type: "text"; value: string }
  | { name: string; type: "integer"; value: number };

export function setupEndpoints(def: Definition) {
  for (const entrypoint of def.entrypoints) {
    for (const endpoint of entrypoint.endpoints) {
      switch (endpoint.kind) {
        case "get": {
          const { path, params } = buildEndpointPath(endpoint);
          app.get(path, (req: Request, res) => {
            // collect path params
            const vars = params.map((p) => validatePathParam(p, req.params[p.name]));

            // get targets
            const q = queryableFromEndpointTargets(
              def,
              endpoint.targets,
              selectToSelectable(endpoint.response),
              "single"
            );
          });
        }
      }
    }
  }
}

/**
 * Here we go again... Returning a query/able, whatever it may be.


from Org.public_repos.issues
filter
      Org.slug is ...
  and public_repos.slug is ...
  and     
*/
