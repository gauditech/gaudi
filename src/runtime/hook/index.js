/** Returns given value. Can be used for testing. */
module.exports.echo = function ({ value }) {
  return value;
};
