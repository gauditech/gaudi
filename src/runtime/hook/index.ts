/** Concatente all given string parts */
export function concat(...parts: string[]): string {
  return parts.join("");
}
