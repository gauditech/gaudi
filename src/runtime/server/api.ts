import path from "path";

import { Express, NextFunction, Request, Response, static as staticHandler } from "express";
import { OpenAPIV3 } from "openapi-types";
import { serve, setup } from "swagger-ui-express";

import { buildOpenAPI } from "@src/builder/openAPI";
import { saveOutputFile } from "@src/common/utils";
import { buildEntrypoints } from "@src/runtime/server/admin";
import { buildEndpoints as buildAuthEndpoints } from "@src/runtime/server/authentication";
import { getAppContext } from "@src/runtime/server/context";
import { buildEndpointConfig, registerServerEndpoint } from "@src/runtime/server/endpoints";
import { EndpointConfig } from "@src/runtime/server/types";
import { Definition, EntrypointDef } from "@src/types/definition";

/** Create endpoint handlers, OpenAPI specs and attach them to server instance */
export function setupServerApis(definition: Definition, app: Express) {
  const config = getAppContext(app).config;

  // --- static folder (eg. for API specs)
  const specOutputFolder = path.join(config.outputFolder, "api-spec");
  app.use("/api-spec", staticHandler(specOutputFolder));

  // --- definition API (aka the "blueprint")
  const definitionEntrypoints = definition.entrypoints;
  const definitionEndpointConfigs = buildEndpointConfig(definition, definitionEntrypoints);
  setupEntrypointApi(
    "api",
    definition,
    definitionEntrypoints,
    definitionEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- admin API
  const adminEntrypoints = buildEntrypoints(definition);
  const adminEndpointConfigs = buildEndpointConfig(definition, adminEntrypoints);
  setupEntrypointApi(
    "api-admin",
    definition,
    adminEntrypoints,
    adminEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- authentication API
  const authConfigs = buildAuthEndpoints();
  setupConfigEndpoints("api", authConfigs, app);
}

/** Create API endpoints from entrypoints.*/
function setupEntrypointApi(
  name: string,
  definition: Definition,
  entrypoints: EntrypointDef[],
  endpointConfigs: EndpointConfig[],
  app: Express,
  outputFolder: string
) {
  const basePath = `/${name}`;
  const specFile = path.join(outputFolder, `${name}.openapi.json`);

  endpointConfigs.forEach((epc) => registerServerEndpoint(app, epc, basePath));

  setupEntrypointApiSpec(definition, entrypoints, app, basePath, specFile);
}

/** Create API OpenAPI spec from entrpyoints */
function setupEntrypointApiSpec(
  definition: Definition,
  entrypoints: EntrypointDef[],
  app: Express,
  basePath: string,
  outputFile: string
) {
  const openApi = buildOpenAPI(
    { models: definition.models, entrypoints, populators: [] },
    basePath
  );

  saveOutputFile(outputFile, JSON.stringify(openApi, undefined, 2));

  setupEntrypointApiSwagger(openApi, app, basePath);
}

function setupEntrypointApiSwagger(
  openApiDocument: OpenAPIV3.Document,
  app: Express,
  apiPath: string
) {
  const swaggerPath = `${apiPath}-docs`;

  app.use(swaggerPath, serve, (_req: Request, _resp: Response, _next: NextFunction) =>
    setup(openApiDocument)(_req, _resp, _next)
  );
}

/** Setup API endpoints from endpoint configs */
function setupConfigEndpoints(name: string, configs: EndpointConfig[], app: Express) {
  const basePath = `/${name}`;

  configs.forEach((epc) => {
    registerServerEndpoint(app, epc, basePath);
  });

  // NOTE: we cannot currently create OpenAPI spec from endpoint configs
}
