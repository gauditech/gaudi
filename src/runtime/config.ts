import fs from "fs";

import { initializeContext } from "@src/runtime/server/context";
import { createDbConn } from "@src/runtime/server/dbConn";
import { Definition } from "@src/types/definition";

export type RuntimeConfig = {
  /** Runtime server host name */
  host: string;
  /** Runtime server port number */
  port: number;
  /** Path to generated definition.json file. */
  definitionPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** DB connection URL */
  dbConnUrl: string;
  /** DB default schema */
  dbSchema?: string;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(): RuntimeConfig {
  const host = process.env.GAUDI_RUNTIME_SERVER_HOST ?? "127.0.0.1";
  const port =
    process.env.GAUDI_RUNTIME_SERVER_PORT != null
      ? parseInt(process.env.GAUDI_RUNTIME_SERVER_PORT, 10)
      : 3001;
  const definitionPath = process.env.GAUDI_RUNTIME_DEFINITION_PATH || "definition.json";
  const outputFolder = process.env.GAUDI_RUNTIME_OUTPUT_PATH || ".";

  const dbConnUrl = process.env.GAUDI_DATABASE_URL || "";
  const dbSchema = process.env.GAUDI_DATABASE_SCHEMA;

  return { host, port, definitionPath, outputFolder, dbConnUrl, dbSchema };
}

/** Load definition file and return it's content */
export function loadDefinition(definitionPath: string): Definition {
  // --- read input file
  if (!fs.existsSync(definitionPath)) {
    throw new Error(`Definition file not found: "${definitionPath}"`);
  }
  const definitionStr = fs.readFileSync(definitionPath).toString("utf-8");
  return JSON.parse(definitionStr);
}

export function createAppContext(config: RuntimeConfig) {
  initializeContext({
    dbConn: createDbConn(config.dbConnUrl, {
      schema: config.dbSchema,
    }),
    config: config,
  });
}
