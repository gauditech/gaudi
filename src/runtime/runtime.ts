// import this file only with relative path because this file actually configures path aliases (eg @src, ...)
import "../common/setupAliases";

import { loadDefinition, readConfig } from "@src/runtime/config";
import { importHooks } from "@src/runtime/hooks";
import { createServer } from "@src/runtime/server/server";

// read environment
const config = readConfig();

const definition = loadDefinition(config.definitionPath);

importHooks(config);

// start server
createServer(definition, config);
