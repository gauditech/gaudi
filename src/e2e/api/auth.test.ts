import path from "path";

import _ from "lodash";
import request from "supertest";

import { createApiTestSetup, loadBlueprint } from "@src/e2e/api/setup";
import { readConfig } from "@src/runtime/config";

// these tests last longer than default 5s timeout so this seems to help
jest.setTimeout(10000);

describe("Auth", () => {
  const config = readConfig(path.join(__dirname, "api.test.env"));

  const { getServer, setup, destroy } = createApiTestSetup(
    config,
    loadBlueprint(path.join(__dirname, "auth.gaudi")),
    [
      {
        model: "AuthUser",
        data: [
          {
            id: 1,
            name: "First",
            username: "first",
            password: "$2b$10$TQpDb3kHc3yLLwtQlM3Rve/ZhUPF7ZZ3WdZ90OxygOCmb7YH.AT86",
          },
          {
            id: 2,
            name: "Second",
            username: "second",
            password: "$2b$10$TQpDb3kHc3yLLwtQlM3Rve/ZhUPF7ZZ3WdZ90OxygOCmb7YH.AT86",
          },
        ],
      },
      {
        model: "AuthUserAccessToken",
        data: [
          {
            authUser_id: 1,
            token: "6Jty8G-HtB9CmB9xqRkJ3Z9LY5_or7pACnAQ6dERc1U",
            expiryDate: `${Date.now() + 1 * 60 * 60 * 1000}`,
          },
          {
            authUser_id: 2,
            token: "FwExbO7sVwf95pI3F3qWSpkANE4aeoNiI0pogqiMcfQ",
            expiryDate: `${Date.now() + 1 * 60 * 60 * 1000}`,
          },
        ],
      },
      {
        model: "Operator",
        data: [
          {
            user_id: 1,
          },
        ],
      },
      {
        model: "Box",
        data: [
          { owner_id: 1, name: "public", is_public: true },
          { owner_id: 1, name: "private", is_public: false },
        ],
      },
      {
        model: "Item",
        data: [
          { box_id: 1, name: "public", is_public: true },
          { box_id: 1, name: "private", is_public: false },
          { box_id: 2, name: "public", is_public: true },
          { box_id: 2, name: "private", is_public: false },
        ],
      },
    ]
  );

  async function loginTestUser() {
    const loginResponse = await request(getServer())
      .post("/auth_user/login")
      .send({ username: "first", password: "1234" });
    return loginResponse.body.token;
  }

  async function loginTestUser2() {
    const loginResponse = await request(getServer())
      .post("/auth_user/login")
      .send({ username: "second", password: "1234" });
    return loginResponse.body.token;
  }

  describe("Login and Logout", () => {
    beforeAll(async () => {
      await setup();
    });
    afterAll(async () => {
      await destroy();
    });

    it("Login and Logout successfully", async () => {
      const listResponse1 = await request(getServer()).get("/box");
      expect(listResponse1.statusCode).toBe(401);

      const loginResponse = await request(getServer())
        .post("/auth_user/login")
        .send({ username: "first", password: "1234" });
      expect(loginResponse.statusCode).toBe(200);

      const token = loginResponse.body.token;
      expect(token?.length).toBe(43);

      const listResponse2 = await request(getServer())
        .get("/box")
        .set("Authorization", "bearer " + token);
      expect(listResponse2.statusCode).toBe(200);

      const logoutResponse = await request(getServer())
        .post("/auth_user/logout")
        .set("Authorization", "bearer " + token);
      expect(logoutResponse.statusCode).toBe(204);

      const listResponse3 = await request(getServer())
        .get("/box")
        .set("Authorization", "bearer " + token);
      expect(listResponse3.statusCode).toBe(401);
    });

    it("Wrong Login password", async () => {
      const loginResponse = await request(getServer())
        .post("/auth_user/login")
        .send({ username: "first", password: "wrong password" });
      expect(loginResponse.statusCode).toBe(401);
    });

    it("Wrong Login username", async () => {
      const loginResponse = await request(getServer())
        .post("/auth_user/login")
        .send({ username: "wrong username", password: "1234" });
      expect(loginResponse.statusCode).toBe(401);
    });
    it("Success public", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/public")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(200);
    });

    it("Success private owned", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/private")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(200);
    });

    it("Fail private", async () => {
      const token = await loginTestUser2();
      const getResponse = await request(getServer())
        .get("/box/private")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(401);
    });

    it("Fail private no auth", async () => {
      const getResponse = await request(getServer()).get("/box/private");
      expect(getResponse.statusCode).toBe(401);
    });

    it("Success public > public", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/public/items/public")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(200);
    });

    it("Fail private > private", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/private/items/private")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(401);
    });

    it("Fail public > private", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/public/items/private")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(401);
    });

    it("Success private owned > public", async () => {
      const token = await loginTestUser();
      const getResponse = await request(getServer())
        .get("/box/private/items/public")
        .set("Authorization", "bearer " + token);
      expect(getResponse.statusCode).toBe(200);
    });

    it("Success create box", async () => {
      const token = await loginTestUser2();
      const getResponse = await request(getServer())
        .post("/box")
        .set("Authorization", "bearer " + token)
        .send({ name: "new box", is_public: false });
      expect(getResponse.statusCode).toBe(200);
    });

    it("Fail create box not logged in", async () => {
      const getResponse = await request(getServer())
        .post("/box")
        .send({ name: "another box", is_public: false });
      expect(getResponse.statusCode).toBe(401);
    });
  });
});
