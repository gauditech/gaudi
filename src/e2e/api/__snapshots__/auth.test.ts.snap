// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Login and Logout should fail when creating user with existing username 1`] = `
{
  "code": "ERROR_CODE_VALIDATION",
  "data": {
    "errors": [
      "Username "somename2@example.com" already exists",
    ],
    "inner": [
      {
        "errors": [
          "Username "somename2@example.com" already exists",
        ],
        "inner": [],
        "message": "Username "somename2@example.com" already exists",
        "name": "ValidationError",
        "params": {
          "originalValue": "somename2@example.com",
          "path": "username",
          "value": "somename2@example.com",
        },
        "path": "username",
        "type": "",
        "value": "somename2@example.com",
      },
    ],
    "value": {
      "username": "somename2@example.com",
    },
  },
  "message": "Validation error",
}
`;

exports[`Auth Login and Logout should fail when creating user with invalid parameters 1`] = `
{
  "code": "ERROR_CODE_VALIDATION",
  "data": {
    "errors": [
      "username must be at least 8 characters",
      "name must be at least 1 characters",
      "password must be at least 8 characters",
      "userProfile.displayName must be defined",
    ],
    "inner": [
      {
        "errors": [
          "username must be at least 8 characters",
        ],
        "inner": [],
        "message": "username must be at least 8 characters",
        "name": "ValidationError",
        "params": {
          "min": 8,
          "originalValue": "",
          "path": "username",
          "value": "",
        },
        "path": "username",
        "type": "min",
        "value": "",
      },
      {
        "errors": [
          "name must be at least 1 characters",
        ],
        "inner": [],
        "message": "name must be at least 1 characters",
        "name": "ValidationError",
        "params": {
          "min": 1,
          "originalValue": "",
          "path": "name",
          "value": "",
        },
        "path": "name",
        "type": "min",
        "value": "",
      },
      {
        "errors": [
          "password must be at least 8 characters",
        ],
        "inner": [],
        "message": "password must be at least 8 characters",
        "name": "ValidationError",
        "params": {
          "min": 8,
          "originalValue": "",
          "path": "password",
          "value": "",
        },
        "path": "password",
        "type": "min",
        "value": "",
      },
      {
        "errors": [
          "userProfile.displayName must be defined",
        ],
        "inner": [],
        "message": "userProfile.displayName must be defined",
        "name": "ValidationError",
        "params": {
          "path": "userProfile.displayName",
        },
        "path": "userProfile.displayName",
        "type": "defined",
      },
    ],
    "value": {
      "name": "",
      "password": "",
      "userProfile": {},
      "username": "",
    },
  },
  "message": "Validation error",
}
`;
